File: eel-core/src/main/scala/io/eels/component/jdbc/JdbcSource.scala
package io.eels.component.jdbc

import java.sql.{Connection, DriverManager, PreparedStatement}

import com.sksamuel.exts.Logging
import com.sksamuel.exts.io.Using
import com.sksamuel.exts.metrics.Timed
import io.eels.component.jdbc.dialect.{GenericJdbcDialect, JdbcDialect}
import io.eels.{Part, Source}
import io.eels.schema.StructType

object JdbcSource {
  def apply(url: String, query: String): JdbcSource = JdbcSource(() => DriverManager.getConnection(url), query)
}

case class JdbcSource(connFn: () => Connection,
                      query: String,
                      bindFn: (PreparedStatement) => Unit = stmt => (),
                      fetchSize: Int = 100,
                      providedSchema: Option[StructType] = None,
                      providedDialect: Option[JdbcDialect] = None,
                      partitionStrategy: JdbcPartitionStrategy = SinglePartitionStrategy)
  extends Source with JdbcPrimitives with Logging with Using with Timed {

  override lazy val schema: StructType = providedSchema.getOrElse(fetchSchema())

  def withBind(bind: (PreparedStatement) => Unit): JdbcSource = copy(bindFn = bind)
  def withFetchSize(fetchSize: Int): JdbcSource = copy(fetchSize = fetchSize)
  def withProvidedSchema(schema: StructType): JdbcSource = copy(providedSchema = Option(schema))
  def withProvidedDialect(dialect: JdbcDialect): JdbcSource = copy(providedDialect = Option(dialect))
  def withPartitionStrategy(strategy: JdbcPartitionStrategy): JdbcSource = copy(partitionStrategy = strategy)

  private def dialect(): JdbcDialect = providedDialect.getOrElse(new GenericJdbcDialect())

  override def parts(): Seq[Part] = partitionStrategy.parts(connFn, query, bindFn, fetchSize, dialect())

  def fetchSchema(): StructType = {
    using(connFn()) { conn =>
      val schemaQuery = s"SELECT * FROM ($query) tmp WHERE 1=0"
      using(conn.prepareStatement(schemaQuery)) { stmt =>

        stmt.setFetchSize(fetchSize)
        bindFn(stmt)

        val rs = timed(s"Executing query $query") {
          stmt.executeQuery()
        }

        val schema = schemaFor(dialect(), rs)
        rs.close()
        schema
      }
    }
  }
}

File: eel-core/src/main/kotlin/io/eels/component/jdbc/JdbcSource.kt
package io.eels.component.jdbc

import io.eels.RowListener
import io.eels.Source
import io.eels.component.Part
import io.eels.component.Using
import io.eels.schema.Schema
import io.eels.util.Logging
import io.eels.util.Option
import io.eels.util.Timed
import io.eels.util.getOrElse
import java.sql.Connection
import java.sql.DriverManager
import java.sql.PreparedStatement

data class JdbcSource @JvmOverloads constructor(val connFn: () -> Connection,
                                                val query: String,
                                                val bind: (PreparedStatement) -> Any = {},
                                                val fetchSize: Int = 100,
                                                val providedSchema: Option<Schema> = Option.None,
                                                val providedDialect: Option<JdbcDialect> = Option.None,
                                                val bucketing: Option<Bucketing> = Option.None,
                                                val listener: RowListener = RowListener.Noop) : Source, JdbcPrimitives, Logging, Using, Timed {

  constructor(url: String, query: String) : this({ DriverManager.getConnection(url) }, query)

  override fun schema(): Schema = providedSchema.getOrElse { fetchSchema() }

  fun withBind(bind: (PreparedStatement) -> Any) = copy(bind = bind)
  fun withFetchSize(fetchSize: Int): JdbcSource = copy(fetchSize = fetchSize)
  fun withProvidedSchema(schema: Schema): JdbcSource = copy(providedSchema = Option(schema))
  fun withProvidedDialect(dialect: JdbcDialect): JdbcSource = copy(providedDialect = Option(dialect))
  fun withListener(listener: RowListener) = copy(listener = listener)

  private fun dialect(): JdbcDialect = providedDialect.getOrElse(GenericJdbcDialect())

  override fun parts(): List<Part> {

    val conn = connFn()
    val stmt = conn.prepareStatement(query)
    stmt.fetchSize = fetchSize
    bind(stmt)

    val rs = timed("Executing query $query") {
      stmt.executeQuery()
    }

    val schema = schemaFor(dialect(), rs)
    val part = ResultsetPart(rs, stmt, conn, schema, listener)
    return listOf(part)
  }

  fun fetchSchema(): Schema {
    return using(connFn()) { conn ->
      using(conn.createStatement()) { stmt ->

        stmt.fetchSize = fetchSize

        val schemaQuery = "SELECT * FROM ($query) tmp WHERE 1=0"
        val rs = timed("Query for schema [$schemaQuery]...") {
          stmt.executeQuery(schemaQuery)
        }

        val schema = schemaFor(dialect(), rs)
        rs.close()
        schema
      }
    }
  }

  data class Bucketing(val columnName: String, val numberOfBuckets: Int)
}